packages/ui/src/views/entity-reference/entity-reference-component.tsx

```diff
@@ -0,0 +1,217 @@
+import React, { useCallback } from 'react'
+
+import { Button, Icon, Input, ScrollArea, SkeletonList } from '@/components'
+import { Root as StackedList, Field as StackedListField, Item as StackedListItem } from '@/components/stacked-list'
+import { cn } from '@utils/cn'
+import { DirectoryType } from '@views/secrets'
+
+// Base properties that all entities must have
+export interface BaseEntityProps {
+  id: string
+  name: string
+}
+
+// Props for rendering a single entity item
+export interface EntityRendererProps<T extends BaseEntityProps> {
+  entity: T
+  isSelected: boolean
+  onSelect: (entity: T) => void
+}
+
+// Props for the scope selector item
+export interface ScopeSelectorProps<S = string> {
+  parentFolder: S
+  onSelect: (scope: S) => void
+}
+
+export interface FolderRendererProps<F = string> {
+  folder: F
+  onSelect: (folder: F) => void
+}
+
+export interface EntityReferenceProps<T extends BaseEntityProps, S = string, F = string> {
+  // Data
+  entities: T[]
+  selectedEntity: T | null
+  parentFolder: S | null
+  childFolder: F | null
+
+  // Callbacks
+  onSelectEntity: (entity: T) => void
+  onScopeChange: (direction: DirectoryType) => void
+
+  // UI Configuration
+  showFilter?: boolean
+
+  // Custom renderers
+  renderEntity?: (props: EntityRendererProps<T>) => React.ReactNode
+  renderScopeSelector?: (props: ScopeSelectorProps<S>) => React.ReactNode
+  renderFolder?: (props: FolderRendererProps<F>) => React.ReactNode
+  isLoading?: boolean
+}
+
+export function EntityReference<T extends BaseEntityProps, S = string, F = string>({
+  // Data
+  entities,
+  selectedEntity,
+  parentFolder,
+  childFolder,
+
+  // Callbacks
+  onSelectEntity,
+  onScopeChange,
+
+  showFilter = true,
+
+  // Custom renderers
+  renderEntity,
+  renderScopeSelector,
+  renderFolder,
+  isLoading = false
+}: EntityReferenceProps<T, S, F>): JSX.Element {
+  const handleSelectEntity = useCallback(
+    (entity: T) => {
+      onSelectEntity?.(entity)
+    },
+    [onSelectEntity]
+  )
+
+  const handleScopeChange = useCallback(
+    (direction: DirectoryType) => {
+      onScopeChange?.(direction)
+    },
+    [onScopeChange]
+  )
+
+  const defaultEntityRenderer = ({ entity, isSelected, onSelect }: EntityRendererProps<T>) => {
+    return (
+      <StackedListItem
+        onClick={() => onSelect?.(entity)}
+        className={cn(isSelected && 'bg-background-4')}
```

Comment:
generally better to avoid the `&&` syntax.
```suggestion
        className={cn({ 'bg-background-4': isSelected })}
```